bench/run.cpp

```diff
@@ -0,0 +1,35 @@
+#include <benchmark/benchmark.h>
+#include <hello_world.hpp>
+
+static void BM_expensive(benchmark::State& state) // NOLINT google-runtime-references
+{
+    if (state.thread_index == 0)
+    {
+    }
+    while (state.KeepRunning())
+    {
+        std::string value = hello_world::expensive(static_cast<std::size_t>(state.range(0)));
```

Comment:
You need to be careful here - as it is the compiler could detect this as dead code and completely optimize it away. Especially since it sees all the code and implementation (because it's all in headers)

You may need to defeat the optimizer here: `escape(value.data())` with

https://gist.github.com/daniel-j-h/0a1c1414a68d4a67d4b4

i think Chandler Carruth has some talks about these issues.

Reply:
Thanks @daniel-j-h , I gave this a try but getting some compile errors. Perhaps I need a particular flag?

Changes:
```c++
static void escape(void* p) 
{
  asm volatile("" : : "g"(p) : "memory");
}

static void clobber() {
  asm volatile("" : : : "memory");
}

static void BM_expensive(benchmark::State& state) // NOLINT google-runtime-references
{
    if (state.thread_index == 0)
    {
    }
    while (state.KeepRunning())
    {
        std::string value = hello_world::expensive(static_cast<std::size_t>(state.range(0)));
        escape(value.data());
        clobber();
    }
    if (state.thread_index == 0)
    {
        // Teardown code here.
    }
}
```

And getting the following error:
```
/Users/carol/dev/hpp-skel/bench/run.cpp:21:9: error: no matching function for call to 'escape'
        escape(value.data());
        ^~~~~~
/Users/carol/dev/hpp-skel/bench/run.cpp:4:13: note: candidate function not viable: no known conversion from 'const value_type *' (aka 'const char *') to
      'void *' for 1st argument
static void escape(void* p) 
            ^
1 error generated.
```

Reply:
@daniel-j-h should @GretaCB  use `benchmark::DoNotOptimize()` or is your `escape/clobber` preferable?

Reply:
Using bench's methods compiled fine, though I couldnt really tell a difference between the benchmark results before/after. Should I see a diff? Also, is optimization a concern for the exclaim benchmark as well?

Reply:
@GretaCB you could put a print statement inside `hello_world::expensive` to see if it is getting called before and after your addition of `benchmark::DoNotOptimize()`. If it prints out both with and without it, my guess is that the code before adding `benchmark::DoNotOptimize()` was working correctly - in the sense that it was not being optimized away and we were actually testing what we thing we were testing. But even if this is the case then @daniel-j-h's point stands that using `benchmark::DoNotOptimize()` is important because each compiler is different and they all try to aggresively optimize. Having `benchmark::DoNotOptimize()` is both a good thing for self-documenting that we expect that code to run and for ensuring it does. (Although even that cannot prevent the compiler from removing code in some cases: https://github.com/google/benchmark/issues/242 so we need to watch out for that)

Reply:
BTW @GretaCB after reading google/benchmark#242 the other day I also was scared the benchmark over at https://github.com/mapbox/vector-tile/ was not working as I expected and that some code might be being removed without me knowing it. However I added https://github.com/mapbox/vector-tile/commit/09245e36987ec52e37f47998dd3beef51ba26c0a#diff-1d28d258250876fb7dde22a17436ef9c to be sure this was not happening and I found that the benchmark speed was the same. Because those functions that returned variables that we unused are very expensive, I feel pretty confident that the code was infact not being optimized away before I made my change in https://github.com/mapbox/vector-tile/commit/09245e36987ec52e37f47998dd3beef51ba26c0a#diff-1d28d258250876fb7dde22a17436ef9c

Reply:
Ah nice I wasn't aware of `benchmark::DoNotOptimize` - looks perfect for this use case!